Add function add_xmlid to OpenUpgrade API

alexis-via opened the pull request.

Commit:
Add function add_xmlid to OpenUpgrade API

StefanRijnhart commented on openerp/openupgrade/openupgrade.py:
```diff
@@ -180,6 +181,38 @@ def rename_xmlids(cr, xmlids_spec):
+    cr.execute(
+        "SELECT id, res_id, noupdate FROM ir_model_data WHERE module=%s AND "
+        "name=%s AND module=%s",
+        (module, xmlid, model, res_id))
```
You have 2x module=%s, the last one looks like it should be model=%s. You can leave out the res_id argument.

StefanRijnhart commented on openerp/openupgrade/openupgrade.py:
```diff
@@ -180,6 +181,38 @@ def rename_xmlids(cr, xmlids_spec):
+        (module, xmlid, model, res_id))
+    already_exists = cr.fetchone()
+    if already_exists:
+        if already_exists[1] != res_id or already_exists[2] != noupdate:
```
Maybe don't update an existing entry, or make it conditional with an extra function argument 'update', defaulting to False. I would not expect an unconditional update like this.

StefanRijnhart commented on openerp/openupgrade/openupgrade.py:
```diff
@@ -180,6 +181,38 @@ def rename_xmlids(cr, xmlids_spec):
                      "WHERE module = %s and name = %s")
             logged_query(cr, query, tuple(new.split('.') + old.split('.')))
 
+def add_xmlid(cr, module, xmlid, model, res_id, noupdate):
```
Could make noupdate a keyword argument defaulting to False

StefanRijnhart commented on openerp/openupgrade/openupgrade.py:
```diff
@@ -180,6 +181,38 @@ def rename_xmlids(cr, xmlids_spec):
+            "INSERT INTO ir_model_data (create_uid, create_date, "
+            "write_uid, write_date, date_init, date_update, noupdate, "
+            "name, module, model, res_id) "
+            "VALUES (%d, (now() at time zone 'UTC'), %d, "
```
From the psycopg docs: the variables placeholder must always be a %s, even if a different placeholder (such as a %d for integers or %f for floats) may look more appropriate

StefanRijnhart commented:
Thanks BTW, looks good in spite of my nags!

bwrsandman commented:
Looks good.
Just a note: for the inserts and database manipulations, you might want to consider using `logged_query()` for better debugging.

Commit:
Take into account the remarks of Stefan :
- noupdate defaults to False
- don't use %d in pscopg
- don't update the existing entry in ir_model_data

bwrsandman commented on openerp/openupgrade/openupgrade.py:
```diff
@@ -180,6 +181,36 @@ def rename_xmlids(cr, xmlids_spec):
+    if already_exists:
+        return False
+    else:
+        cr.execute(
```
`logged_query()` would be better here so we can trace SQL calls

Example:

```
logged_query(cr, """
INSERT INTO ir_model_data (create_uid, create_date,
write_uid, write_date, date_init, date_update, noupdate,
name, module, model, res_id)
VALUES (%s, (now() at time zone 'UTC'), %s,
(now() at time zone 'UTC'), (now() at time zone 'UTC'),
(now() at time zone 'UTC'), %s, %s, %s, %s, %s)
""", (SUPERUSER_ID, SUPERUSER_ID, noupdate,
       xmlid, module, model, res_id))
```

StefanRijnhart commented:
Thanks for the changes. I think not having logged_query is fine for one off inserts.
:+1:

StefanRijnhart commented:
Just for the record, after a short discussion we agreed to have logged_query in such cases as well so :-1: until then

Commit:
Use logged_query() instead of cr.execute()

paulius-sladkevicius commented:
:+1:

bwrsandman commented:
Looks good, just a few pep8 errors.
You can check them on the link in "Details"

Commit:
PEP8 stuff

StefanRijnhart commented:
Tnx, pep-bot is green ;-) :+1:
